design/2022.exponential-backoff.md

```diff
@@ -0,0 +1,126 @@
+  renewalTime: <timestamp>
+  attemptsForRevision: 3
+  lastFailureTime: <timestamp> # Last failed issuance (i.e when a `CertificateRequest` failed)
+  revision: 19 # The latest succeeded revision. Currently we are attempting revision 20
+  ```
+
+2. A `Certificate` where the latest issuance for revision 20 succeeded and no issuances are being attempted now:
+status:
+  conditions:
+  - lastTransitionTime: <timestamp>
+    message: <message>
+    observedGeneration: 1
+    reason: Ready
+    status: "True"
+    type: Ready
+  notAfter: <timestamp>
+  notBefore: <timestamp>
+  renewalTime: <timestamp>
+  revision: 20 
+
+
+A `Certificate`'s revision is currently recorded using `status.revision` field of a `Certificate` and a `cert-manager.io/certificate-revision` annotation on a `CertificateRequest`. Each time a renewal is attempted (either because it is time to renew or renewal has been triggered by other means such as running `cmctl renew`), `CertificateRequest` for the renewal will have a revision of 1 + current_revision where current_revision is the revision currently on `Certificate`'s status (assumed to be 0 for a newly created `Certificate`). In case of a successful issuance, current_revision is bumped by 1. In case of failed issuance, current revision remains the same, meaning that the next `CertificateRequest` that gets created will have the same revision value as the `CertificateRequest` that just failed.
+
+#### Scenarios:
+
+##### Forcing re-issuance
+
+A successful issuance will be the only way how the backoff period gets reset. Deleting `CertificateRequest` will not affect the backoff period as it is determined by looking at the status of the `Certificate`. This is similar to the current behaviour where, when a `CertificateRequest` is _not_ found, but `status.LastFailureTime` is set on a `Certificate`, we fall back to backing off as opposed to falling back to issuing now, see [trigger-controller#shouldbackoffReissuingOnFailure](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/trigger/trigger_controller.go#L216-L218)
+However, users will be able to run `cmctl renew`, which will trigger an immediate re-issuance regardless of the backoff period. If this issuance succeeds, `status.IssuanceAttemptsForRevision` will be removed thus resetting the backoff period.
+
+
+
+A new status field will be added on the `Certificate` resource to help with calculating how many times we have failed during _this_ revision.
+
+### Non-goals
+
+- Make the backoff period configurable (extra complexity). For context Kubernetes pod backoff crashloopbackoff period is _not_ configurable (although it is a very demanded feature [k/k#57291](https://github.com/kubernetes/kubernetes/issues/57291))
+
+### Must not
+- Break the current abstractions of Certificate, CertificateRequest, Order etc. For example,`CertificateRequest` represents a single attempt to issue a certificate that either succeeded or not if it's in terminal state
```

Comment:
does this mean "Break compatability", i.e. it must not break compatability, up- and downgrading can be done without extra steps.

Reply:
No, this is more abstract, an example breaking of abstraction would be, I think, making `CertificateRequest` somehow 'aware' of the next or previous issuances, for example a field that would somehow encode the delay for this particular issuance. It's related to what could be alternative ways to do this, but maybe actually doesn't make that much sense in the context of this doc

Reply:
Thanks for clarifying. 

**Should** we have a "must not" item specifically for compatibility, or is that implied?

Reply:
Thanks, it's a very good point 👍🏼 
I've added an 'Upgrading' section that should cover this